src/ng/http.js

```diff
@@ -1135,29 +1135,86 @@ function $HttpProvider() {
     }
 
 
-    function buildUrl(url, params) {
-      if (!params) return url;
-      var parts = [];
-      forEachSorted(params, function(value, key) {
-        if (value === null || isUndefined(value)) return;
-        if (!isArray(value)) value = [value];
-
-        forEach(value, function(v) {
-          if (isObject(v)) {
-            if (isDate(v)) {
-              v = v.toISOString();
-            } else {
-              v = toJson(v);
-            }
-          }
-          parts.push(encodeUriQuery(key) + '=' +
-                     encodeUriQuery(v));
-        });
-      });
-      if (parts.length > 0) {
-        url += ((url.indexOf('?') == -1) ? '?' : '&') + parts.join('&');
+    function buildUrl(url, params, mode) {
+      params = $httpUrlParams.serialize(params, mode);
+      if (params) {
+        url += ((url.indexOf('?') == -1) ? '?' : '&') + params;
       }
       return url;
     }
   }];
 }
+
+/**
+ * @ngdoc provider
+ * @name $httpUrlParamsProvider
+ * @description
+ * Use `$httpUrlParamsProvider` to change the default behavior of the {@link ng.$httpUrlParams $httpUrlParams} service.
+ * */
+function $HttpUrlParamsProvider() {
+
+  var paramSerializers = createMap();
+
+  function serializeParams(params, addArrayMarker) {
+    var parts = [];
+
+    if (!params) return '';
+
+    forEachSorted(params, function(value, key) {
+      if (value === null || isUndefined(value)) return;
+      if (!isArray(value)) value = [value];
+
+      forEach(value, function(v) {
+        if (isObject(v)) {
+          v = isDate(v) ? v.toISOString() : toJson(v);
+        }
+        parts.push(encodeUriQuery(key) + (addArrayMarker ? '[]' : '') + '=' + encodeUriQuery(v));
+      });
+    });
+
+    return parts.join('&');
+  }
+
+
+  this.defaultMode = 'traditional';
+
+
+  this.registerSerializer = function registerSerializer(mode, serFn) {
+    paramSerializers[mode] = serFn;
+  };
+
+  /**
+  * @ngdoc service
+  * @name $httpUrlParams
+  *
+  * @description
+  * The `$httpUrlParams` service is responsible for serializing request params
+  * (expressed as a JavaScript object) to a string.
+  * It abstracts the way in which request params are transformed, grouped, encoded etc.
+  *
+  * Normally you wouldn't use this service directly in the application's code but rather override this service
+  * to enable custom serialization schemas for URL parameters.
+  *
+  * The `$httpUrlParams` service comes handy while unit testing with {@link ngMock.$httpBackend $httpBackend mock},
+  * as one can inject `$httpUrlParams` into a test and serialize URL params as {@link ng.$http $http} service.
+  */
+  this.$get = function() {
+
+    var provider = this;
+    var HttpUrlParams = {};
+
+    HttpUrlParams.serialize = function(params, mode) {
+      return paramSerializers[lowercase(mode) || provider.defaultMode](params);
```

Comment:
If we decide to throw, it would be better to throw a more meaningful error (instead of `undefined is not a function`).
I would also consider logging the incident and falling back to the default mode.

BTW, you are using `lowercase` here, but not when registering. They should be consistent (and probably documented).